### PR TITLE
Pytestless doctests

### PR DESCRIPTION
**Pull request by pp-mo**
Pursuing an idea that we might create a better test-runner than pytest, for ourselves

### CODE REVIEW

File: tools/run_doctests.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+
+    return result
+
+
+def list_filepaths_recursive(
+    file_path: str, exclude_matches: list[str] = []
+) -> list[Path]:
+    """Expand globs to a list of filepaths.
+
+    Also filter with exclude controls.
```
Comment (pp-mo) [inline tools/run_doctests.py:70]: I think the "private" concept does _not_ apply to documentation files :  Explain this?

File: tools/run_doctests.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    """
+    actual_paths: list[Path] = []
+    segments = file_path.split("/")
+    i_wilds = [
```
Comment (pp-mo) [inline tools/run_doctests.py:74]: TODO: document the intention.

In order to resolve a globbed path like `dir/subdir/**/*AB_test.extension`, we need to split the root path of the search from the wildcarded parts, to do `Path("dir/subdir").glob("**/*AB_test.extension")`
So here we are finding which path segments contain glob characters, and splitting at the first of those.
This probably is still a bit crude 
  - e.g. can't handle escaped glob chars
  - e.g. maybe doesn't work (needs testing) if the globbed part or unglobbed parts are empty

**Comment by pp-mo** [comment]
## What is still missing?
Apart from the above, this is still only "trial code".
Intention is (for now) to **not** put this in a distributable package, but into a [SciTools template](https://github.com/SciTools/.github/blob/main/templates/README.md)

What's really driving that is a desire to use this for Iris too.

Probably, it would still be good to _have_ a distinct repo, for central control + to host testing, 
 \- but that would also **follow the template**
 (or perhaps, ~equivalently, the template follows the primary repo)
 
 What's I can already see is still missing with the function itself ..
   * now complicated enough that it _needs some tests_
   * possible problems with less usual glob forms (in include/target specs) -- see above
     * ideally, do the glob thing properly (e.g. support '**' -- but don't know of a 'standard' implementation for that)
   * support proper globs in exclude patterns ?
   * lack of global options config 
     * possibly should exist in pyproject.toml
     * make it look like a real tool
     * this would seem _especially_ useful for excludes

File: tools/run_doctests.py (hunk 2 above)
Comment (pp-mo) [reply tools/run_doctests.py:74]: **update**:
After some reading, possibly what is really wanted here can be done with a simple glob `recursive=True` ?
See https://stackoverflow.com/a/57644936/2615050

Best I think would be to first get tests working for complicated cases, as is, **_then_** show that the simple glob solution works!

File: tools/run_doctests.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+    # Remove all spaces (think they are never needed).
+    opt_str = opt_str.replace(" ", "")
+    # Split on commas, and split each one on "=" expecting a simple name=val form
+    opts_dict = {}
+    if opt_str:  # N.B. to avoid unexpected behaviour: "".split() --> [""]
+        for setting_str in opt_str.split(","):
+            try:
+                name, val = setting_str.split("=")
+
+                # Detect + translate numeric and boolean values.
```
Comment (pp-mo) [inline tools/run_doctests.py:111]: "numeric" here should probably say "integer".  
Floats etc not needed for anything in these functions

File: tools/run_doctests.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+        if not any(match in str(path) for match in exclude_matches)
+        and not path.name.startswith("_")
+    ]
+    return result
+
+
+def process_options(
+    opt_str: str, paths_are_modules: bool = True
+) -> dict[str, str]:
+    """Convert the "-o/--options" arg into a **kwargs for the doctest function call."""
```
Comment (pp-mo) [inline tools/run_doctests.py:101]: Perhaps worth making the point here that we are 'interpreting' the input to avoid using "exec", on general principles (!)

**Comment by pp-mo** [comment]
Replaced by #196
